tools/src/main/scala/scala/scalanative/compiler/pass/IsLowering.scala

```diff
@@ -8,7 +8,44 @@ import nir._, Inst.Let
 
 /** Translates instance checks to range checks on type ids. */
 class IsLowering(implicit fresh: Fresh, top: Top) extends Pass {
-  override def preInst = {
+
+  override def preInst = eliminateIs
+
+  private def eliminateIs: PartialFunction[Inst, Seq[Inst]] = {
+    case Let(n, Op.Is(_, Val.Zero(_))) =>
+      Seq(Let(n, Op.Copy(Val.False)))
+
+    case isInst @ Let(n, Op.Is(ty, obj)) =>
+      val isNullV = Val.Local(fresh(), Type.Bool)
+      val isIsV   = Val.Local(fresh(), Type.Bool)
+
+      val thenL = fresh()
+      val elseL = fresh()
+      val contL = fresh()
+
+      val thenResultV = Val.Local(fresh(), Type.Bool)
+      val elseResultV = Val.Local(fresh(), Type.Bool)
+
+      Seq(
+          // if
+          Let(isNullV.name,
+              Op.Comp(Comp.Ieq, Type.Ptr, obj, Val.Zero(Type.Ptr))),
+          Inst.If(isNullV, Next(thenL), Next(elseL)),
+          // then
+          Inst.Label(thenL, Seq.empty),
+          Let(thenResultV.name, Op.Copy(Val.False)),
+          Inst.Jump(Next.Label(contL, Seq(thenResultV))),
+          // else
+          Inst.Label(elseL, Seq.empty)) ++
+      doEliminateIs(Let(elseResultV.name, Op.Is(ty, obj))) ++
```

Comment:
I don't think that it will really change much. The current solution looks good to me.